libopae/src/sysobject.c

```diff
@@ -0,0 +1,286 @@
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+#include <sys/types.h>
+#include <sys/stat.h>
+
+#include "common_int.h"
+#include "log_int.h"
+#include "sysfs_int.h"
+#include "types_int.h"
+#include "safe_string/safe_string.h"
+#include <opae/types_enum.h>
+#include <opae/sysobject.h>
+
+#define SYSOBJ_ASSERT_NOT_NULL(var)                                            \
+	do {                                                                   \
+		if (var == NULL) {                                             \
+			FPGA_MSG(#var " is NULL");                             \
+		}                                                              \
+	} while (false);
+
+#define FREE_IF(var)                                                           \
+	do {                                                                   \
+		if (var) {                                                     \
+			free(var);                                             \
+			var = NULL;                                            \
+		}                                                              \
+	} while (false);
+
+
+fpga_result __FPGA_API__ fpgaReadObjectBytes(fpga_token token, const char *key,
+					     uint8_t *buffer, size_t offset,
+					     size_t *len)
+{
+	int fd = -1, fd_stat = 0;
+	char objpath[SYSFS_PATH_MAX];
+	ssize_t count = 0;
+	fpga_result res = FPGA_EXCEPTION;
+	struct stat objstat;
+
+	SYSOBJ_ASSERT_NOT_NULL(token);
+	SYSOBJ_ASSERT_NOT_NULL(key);
+	SYSOBJ_ASSERT_NOT_NULL(len);
+
+	res = cat_token_sysfs_path(objpath, token, key);
+	if (res) {
+		return res;
+	}
+
+	fd_stat = stat(objpath, &objstat);
+	if (fd_stat < 0) {
+		FPGA_ERR("Error with object path: %s", strerror(errno));
+		if (errno == EACCES) {
+			return FPGA_NO_ACCESS;
+		}
+		return FPGA_EXCEPTION;
+	}
+
+	if (!buffer) {
+		*len = objstat.st_size - offset;
+		return FPGA_OK;
+	}
+
+
+	fd = open(objpath, O_RDONLY);
+	if (fd < 0) {
+		return FPGA_NOT_FOUND;
+	}
+
+	count = eintr_read(fd, buffer, *len);
+	if (count < (ssize_t)*len) {
+		if (count < 0) {
+			FPGA_ERR("Error with pread operation: %s",
+				 strerror(errno));
+			res = FPGA_EXCEPTION;
+		} else {
+			FPGA_MSG("Bytes read (%d) is less that requested (%d)",
+				 count, *len);
+			res = count == 0 ? FPGA_EXCEPTION : FPGA_OK;
+		}
+	} else {
+		res = FPGA_OK;
+	}
+	*len = count;
+	close(fd);
+
+	return res;
+}
+
+
+fpga_result fpgaTokenGetObject(fpga_token token, const char *name,
+			       fpga_object *object, int flags)
+{
+	(void)flags;
+	char objpath[SYSFS_PATH_MAX];
+	fpga_result res = FPGA_EXCEPTION;
+
+	SYSOBJ_ASSERT_NOT_NULL(token);
+	SYSOBJ_ASSERT_NOT_NULL(name);
+	res = cat_token_sysfs_path(objpath, token, name);
+	if (res) {
+		return res;
+	}
+
+	return make_sysfs_object(objpath, name, object, flags, NULL);
+}
+
+
+fpga_result fpgaHandleGetObject(fpga_token handle, const char *name,
+				fpga_object *object, int flags)
+{
+	(void)flags;
```

Comment:
Can we use UNUSED_PARAM for consistency?

Reply:
👍